### PR TITLE
Generic and optimized WebSocket API

### PR DESCRIPTION
**Pull request by Centril**
# Description of Changes

Implements https://github.com/clockworklabs/SpacetimeDBPrivate/pull/993 but with some changes.

Fixes https://github.com/clockworklabs/SpacetimeDB/issues/1080.
Fixes https://github.com/clockworklabs/SpacetimeDB/issues/1450.
Fixes https://github.com/clockworklabs/SpacetimeDB/issues/1201.
Closes https://github.com/clockworklabs/SpacetimeDB/pull/1200.

# API and ABI breaking changes

Yes We Can.

### CODE REVIEW

**Comment by bfops** [comment]
Converting to a draft since this is marked draft

File: crates/core/src/client/messages.rs [hunk 1]
```diff
@@ -28,265 +24,201 @@ pub trait ToProtocol {
...
-            let mut encoder = CompressorReader::new(reader, BUFFER_SIZE, COMPRESSION_LEVEL, LG_WIN);
-
-            let mut out = Vec::new();
-            encoder
-                .read_to_end(&mut out)
-                .expect("Failed to Brotli compress `SubscriptionUpdateMessage`");
-            out.into()
+pub fn serialize(msg: impl ToProtocol<Encoded = SwitchedServerMessage>, protocol: Protocol) -> DataMessage {
+    // TODO(centril, perf): here we are allocating buffers only to throw them away eventually.
+    // Consider pooling these allocations so that we reuse them.
```
Comment (coolreader18) [inline crates/core/src/client/messages.rs:34]: Would so love to, but `tungstenite::Message` takes a `Vec<u8>`/`String`, so we'd have to use a different websocket library. good to have this here, though.

File: crates/client-api-messages/examples/get_ws_schema.rs [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+use spacetimedb_client_api_messages::websocket::{BsatnFormat, ClientMessage, ServerMessage};
 use spacetimedb_lib::ser::serde::SerializeWrapper;
 use spacetimedb_lib::{RawModuleDef, RawModuleDefV8};
 
 fn main() -> Result<(), serde_json::Error> {
     let module = RawModuleDefV8::with_builder(|module| {
-        module.add_type::<ClientMessage>();
-        module.add_type::<ServerMessage>();
+        module.add_type::<ClientMessage<Bytes>>();
+        module.add_type::<ServerMessage<BsatnFormat>>();
```
Comment (coolreader18) [inline crates/client-api-messages/examples/get_ws_schema.rs:9]: Maybe this could take a cli arg to choose whether it's json or bsatn?

**Review by coolreader18 — APPROVED** [review]
LGTM, overall!

**Pull request merged**